File: 1930-unique-length-3-palindromic-subsequences/frequency-tracking/cpp/solution.cpp
class Solution {
public:
    int countPalindromicSubsequence(string s) {
        unordered_set<string> res;
        unordered_set<char> left;
        unordered_map<char, int> right;
        for (char ele : s) {
            right[ele]++;
        }

        for (char mid : s) {    // atmost n
            right[mid]--;
            for (char outer : left) {   // atmost 26
                if (right[outer] > 0) {
                    res.insert(string(1, outer) + mid);
                }
            }
            left.insert(mid);
        }
        return res.size();
    }
};

File: 1930-unique-length-3-palindromic-subsequences/frequency-tracking/python/solution.py
class Solution:
    def countPalindromicSubsequence(self, s: str) -> int:
        res = set()
        left = set()
        right = collections.Counter(s)

        for mid in s: # atmost n
            right[mid] -= 1
            for outer in left: # atmost 26
                if right[outer] > 0:
                    res.add((outer, mid))
            left.add(mid)
        
        return len(res)
